### PR TITLE
[ASM] Introduce SecurityReporter for all reporting functions of SecurityCoordinator

### PR DESCRIPTION
**Pull request by anna-git**
## Summary of changes

This PR is just moving code around, to allow reporting security infos on spans without needing a security coordinator. 

Basically we are extracting from `SecurityCoordinator` the reporting functions to the already existing underused `SecurityReporter`, we are nesting core/framework classes inside of `SecurityCoordinator` and `SecurityReporter`.
`SecurityReporter` can function without the `Security` instance, it just needs the span and transport to add info. 

This allows for better separation of concerns and prepares things for ATO as some contexts just need to report additional headers without running the waf necessarily, or already, within the catch (BlockException) inside the `BlockingMiddleware` where we just want to report the result within the exception and don't need any security features

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (6481)](https://github.com/DataDog/dd-trace-dotnet/tree/eafc26eb33868bfe32764f3d5beb5aaecd5f4441)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6481) - mean (68ms)  : 66, 71
     .   : milestone, 68,
    master - mean (68ms)  : 66, 71
     .   : milestone, 68,

    section CallTarget+Inlining+NGEN
    This PR (6481) - mean (977ms)  : 958, 996
     .   : milestone, 977,
    master - mean (978ms)  : 959, 997
     .   : milestone, 978,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6481) - mean (107ms)  : 105, 110
     .   : milestone, 107,
    master - mean (107ms)  : 105, 110
     .   : milestone, 107,

    section CallTarget+Inlining+NGEN
    This PR (6481) - mean (675ms)  : 658, 691
     .   : milestone, 675,
    master - mean (678ms)  : 664, 693
     .   : milestone, 678,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6481) - mean (91ms)  : 89, 93
     .   : milestone, 91,
    master - mean (91ms)  : 89, 94
     .   : milestone, 91,

    section CallTarget+Inlining+NGEN
    This PR (6481) - mean (633ms)  : 615, 652
     .   : milestone, 633,
    master - mean (636ms)  : 618, 653
     .   : milestone, 636,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6481) - mean (196ms)  : 191, 200
     .   : milestone, 196,
    master - mean (193ms)  : 190, 196
     .   : milestone, 193,

    section CallTarget+Inlining+NGEN
    This PR (6481) - mean (1,099ms)  : 1066, 1133
     .   : milestone, 1099,
    master - mean (1,104ms)  : 1076, 1132
     .   : milestone, 1104,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6481) - mean (278ms)  : 274, 283
     .   : milestone, 278,
    master - mean (279ms)  : 275, 283
     .   : milestone, 279,

    section CallTarget+Inlining+NGEN
    This PR (6481) - mean (866ms)  : 833, 900
     .   : milestone, 866,
    master - mean (872ms)  : 851, 893
     .   : milestone, 872,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6481) - mean (267ms)  : 263, 271
     .   : milestone, 267,
    master - mean (267ms)  : 262, 272
     .   : milestone, 267,

    section CallTarget+Inlining+NGEN
    This PR (6481) - mean (849ms)  : 807, 890
     .   : milestone, 849,
    master - mean (851ms)  : 813, 889
     .   : milestone, 851,

```

**Comment by andrewlock** [comment]
## Benchmarks Report for appsec :snail:

Benchmarks for #6481 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9): 
* All benchmarks have **the same** speed
* 1 benchmarks have **fewer** allocations
* 2 benchmarks have **more** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.Asm.AppSecBodyBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`AllCycleSimpleBody`|net6.0|205μs|73.2ns|274ns|2.57| 0|0|188.73&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`AllCycleSimpleBody`|netcoreapp3.1|306μs|153ns|571ns|2.6| 0|0|196.06&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`AllCycleSimpleBody`|net472|269μs|157ns|586ns|35.8| 2.02|0|225.75&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`AllCycleMoreComplexBody`|net6.0|208μs|63.7ns|230ns|2.7| 0|0|192.23&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`AllCycleMoreComplexBody`|netcoreapp3.1|313μs|117ns|421ns|2.65| 0|0|199.47&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`AllCycleMoreComplexBody`|net472|274μs|121ns|437ns|36.3| 2.06|0|229.27&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`ObjectExtractorSimpleBody`|net6.0|138ns|0.125ns|0.468ns|0.00392| 0|0|280&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`ObjectExtractorSimpleBody`|netcoreapp3.1|196ns|0.161ns|0.603ns|0.00369| 0|0|272&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`ObjectExtractorSimpleBody`|net472|166ns|0.172ns|0.667ns|0.0446| 0|0|281&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`ObjectExtractorMoreComplexBody`|net6.0|2.98μs|1.73ns|6.47ns|0.0522| 0|0|3.78&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|3.76μs|2.89ns|11.2ns|0.0506| 0|0|3.69&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`ObjectExtractorMoreComplexBody`|net472|3.59μs|2.59ns|9.71ns|0.602| 0.00541|0|3.8&nbsp;KB|
|#6481|`AllCycleSimpleBody`|net6.0|204μs|125ns|469ns|2.65| 0|0|188.76&nbsp;KB|
|#6481|`AllCycleSimpleBody`|netcoreapp3.1|301μs|116ns|434ns|2.59| 0|0|196.09&nbsp;KB|
|#6481|`AllCycleSimpleBody`|net472|269μs|525ns|1.89μs|35.8| 2|0|225.79&nbsp;KB|
|#6481|`AllCycleMoreComplexBody`|net6.0|211μs|74.2ns|287ns|2.74| 0|0|192.26&nbsp;KB|
|#6481|`AllCycleMoreComplexBody`|netcoreapp3.1|314μs|179ns|668ns|2.66| 0|0|199.5&nbsp;KB|
|#6481|`AllCycleMoreComplexBody`|net472|279μs|82ns|307ns|36.4| 1.97|0|229.3&nbsp;KB|
|#6481|`ObjectExtractorSimpleBody`|net6.0|139ns|0.181ns|0.699ns|0.00398| 0|0|280&nbsp;B|
|#6481|`ObjectExtractorSimpleBody`|netcoreapp3.1|194ns|0.142ns|0.533ns|0.00372| 0|0|272&nbsp;B|
|#6481|`ObjectExtractorSimpleBody`|net472|166ns|0.112ns|0.421ns|0.0446| 0|0|281&nbsp;B|
|#6481|`ObjectExtractorMoreComplexBody`|net6.0|2.9μs|1.81ns|6.54ns|0.0525| 0|0|3.78&nbsp;KB|
|#6481|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|3.8μs|2.03ns|7.31ns|0.0512| 0|0|3.69&nbsp;KB|
|#6481|`ObjectExtractorMoreComplexBody`|net472|3.59μs|1.56ns|5.64ns|0.601| 0.00539|0|3.8&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecEncoderBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`EncodeArgs`|net6.0|37.5μs|15.1ns|56.5ns|0.447| 0|0|32.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`EncodeArgs`|netcoreapp3.1|53.9μs|22.9ns|85.8ns|0.429| 0|0|32.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`EncodeArgs`|net472|66.1μs|46.2ns|179ns|5.14| 0.0664|0|32.5&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`EncodeLegacyArgs`|net6.0|73.7μs|21.6ns|74.7ns|0| 0|0|2.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`EncodeLegacyArgs`|netcoreapp3.1|105μs|170ns|659ns|0| 0|0|2.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`EncodeLegacyArgs`|net472|155μs|91.3ns|354ns|0.308| 0|0|2.15&nbsp;KB|
|#6481|`EncodeArgs`|net6.0|37.6μs|10.6ns|36.8ns|0.448| 0|0|32.4&nbsp;KB|
|#6481|`EncodeArgs`|netcoreapp3.1|55.1μs|37.4ns|140ns|0.436| 0|0|32.4&nbsp;KB|
|#6481|`EncodeArgs`|net472|66.4μs|41.5ns|155ns|5.15| 0.066|0|32.5&nbsp;KB|
|#6481|`EncodeLegacyArgs`|net6.0|79.7μs|28.9ns|112ns|0| 0|0|2.14&nbsp;KB|
|#6481|`EncodeLegacyArgs`|netcoreapp3.1|105μs|145ns|563ns|0| 0|0|2.14&nbsp;KB|
|#6481|`EncodeLegacyArgs`|net472|154μs|96.2ns|373ns|0.308| 0|0|2.15&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecWafBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`RunWafRealisticBenchmark`|net6.0|174μs|92.9ns|360ns|0| 0|0|2.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`RunWafRealisticBenchmark`|netcoreapp3.1|186μs|138ns|535ns|0| 0|0|2.39&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`RunWafRealisticBenchmark`|net472|198μs|133ns|517ns|0.3| 0|0|2.46&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`RunWafRealisticBenchmarkWithAttack`|net6.0|116μs|50.5ns|196ns|0| 0|0|1.47&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`RunWafRealisticBenchmarkWithAttack`|netcoreapp3.1|123μs|178ns|689ns|0| 0|0|1.46&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`RunWafRealisticBenchmarkWithAttack`|net472|132μs|105ns|407ns|0.198| 0|0|1.49&nbsp;KB|
|#6481|`RunWafRealisticBenchmark`|net6.0|174μs|148ns|553ns|0| 0|0|2.44&nbsp;KB|
|#6481|`RunWafRealisticBenchmark`|netcoreapp3.1|188μs|468ns|1.81μs|0| 0|0|2.39&nbsp;KB|
|#6481|`RunWafRealisticBenchmark`|net472|202μs|82.7ns|320ns|0.3| 0|0|2.46&nbsp;KB|
|#6481|`RunWafRealisticBenchmarkWithAttack`|net6.0|116μs|99.3ns|358ns|0| 0|0|1.47&nbsp;KB|
|#6481|`RunWafRealisticBenchmarkWithAttack`|netcoreapp3.1|123μs|232ns|898ns|0| 0|0|1.46&nbsp;KB|
|#6481|`RunWafRealisticBenchmarkWithAttack`|net472|133μs|56.1ns|210ns|0.199| 0|0|1.48&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Iast.StringAspectsBenchmark</strong> - Same speed :heavy_check_mark: More allocations :warning:



#### More allocations :warning: in #6481

|  Benchmark         | Base Allocated | Diff Allocated | Change   | Change % |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatAspectBenchmark&#8209;net6.0 | 255.37 KB | 264.33 KB | 8.96 KB | 3.51%
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatAspectBenchmark&#8209;netcoreapp3.1 | 252.89 KB | 254.9 KB | 2.02 KB | 0.80%



#### Fewer allocations :tada: in #6481

|  Benchmark         | Base Allocated | Diff Allocated | Change   | Change % |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatBenchmark&#8209;net472 | 59.68 KB | 57.34 KB | -2.34 KB | -3.91%

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`StringConcatBenchmark`|net6.0|60.8μs|793ns|7.93μs|0| 0|0|43.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`StringConcatBenchmark`|netcoreapp3.1|59.4μs|679ns|6.72μs|0| 0|0|42.64&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`StringConcatBenchmark`|net472|37.6μs|90.9ns|328ns|0| 0|0|59.68&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`StringConcatAspectBenchmark`|net6.0|314μs|1.47μs|6.06μs|0| 0|0|255.37&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`StringConcatAspectBenchmark`|netcoreapp3.1|336μs|1.54μs|5.55μs|0| 0|0|252.89&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/1ef216bf8d9174aa50ae40923f78e6b338581ca9)|`StringConcatAspectBenchmark`|net472|283μs|6.17μs|60.4μs|0| 0|0|278.53&nbsp;KB|
|#6481|`StringConcatBenchmark`|net6.0|61.6μs|895ns|8.67μs|0| 0|0|43.44&nbsp;KB|
|#6481|`StringConcatBenchmark`|netcoreapp3.1|58.6μs|640ns|6.34μs|0| 0|0|42.64&nbsp;KB|
|#6481|`StringConcatBenchmark`|net472|37.4μs|88.8ns|320ns|0| 0|0|57.34&nbsp;KB|
|#6481|`StringConcatAspectBenchmark`|net6.0|319μs|1.58μs|10.6μs|0| 0|0|264.33&nbsp;KB|
|#6481|`StringConcatAspectBenchmark`|netcoreapp3.1|334μs|1.45μs|5.41μs|0| 0|0|254.9&nbsp;KB|
|#6481|`StringConcatAspectBenchmark`|net472|281μs|5.6μs|55.1μs|0| 0|0|278.53&nbsp;KB|
</details>

**Review by NachoEchevarria — APPROVED** [review]

**Review by daniel-romano-DD — APPROVED** [review]

**Review by e-n-0 — APPROVED** [review]

File: tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs [hunk 1]
```diff
@@ -437,12 +405,7 @@ public Dictionary<string, object> GetBasicRequestArgsForWaf()
             }
         }
 
-        var dict = new Dictionary<string, object>(capacity: 7)
-        {
-            { AddressesConstants.RequestMethod, request.HttpMethod },
-            { AddressesConstants.ResponseStatus, request.RequestContext.HttpContext.Response.StatusCode.ToString() },
-            { AddressesConstants.RequestClientIp, _localRootSpan.GetTag(Tags.HttpClientIp) }
-        };
+        var dict = new Dictionary<string, object>(capacity: 7) { { AddressesConstants.RequestMethod, request.HttpMethod }, { AddressesConstants.ResponseStatus, request.RequestContext.HttpContext.Response.StatusCode.ToString() }, { AddressesConstants.RequestClientIp, _localRootSpan.GetTag(Tags.HttpClientIp) } };
```
Comment (vandonr) [inline tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs:408]: I suppose it's auto-formating that changed this line and no you directly, but I preferred the previous version.

**Review by vandonr — APPROVED** [review]

File: tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs (hunk 1 above)
Comment (anna-git) [reply tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs:408]: I agree, and I'm not sure why it shows like that, cause it's still wrapped in the editor 🤔 
image

**Pull request merged**